More dynamicDowncast<> adoption in layout code

annevk opened the pull request:
#### bac2024d2e9d973d2210f9e06ff5a01272ccccf2
<pre>
More dynamicDowncast&lt;&gt; adoption in layout code
<a href="https://bugs.webkit.org/show_bug.cgi?id=269173">https://bugs.webkit.org/show_bug.cgi?id=269173</a>

Reviewed by Alan Baradlay.

Including a drive-by fix for InspectorAuditAccessibilityObject.

* Source/WebCore/inspector/InspectorAuditAccessibilityObject.cpp:
(WebCore::InspectorAuditAccessibilityObject::getElementsByComputedRole):
(WebCore::InspectorAuditAccessibilityObject::getMouseEventNode):
* Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp:
(WebCore::Layout::partialRangeForDamage):
(WebCore::Layout::isEmptyInlineContent):
* Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
(WebCore::Layout::requiresVisualReordering):
(WebCore::Layout::InlineItemsBuilder::traverseUntilDamaged):
(WebCore::Layout::InlineItemsBuilder::collectInlineItems):
(WebCore::Layout::InlineItemsBuilder::breakAndComputeBidiLevels):
(WebCore::Layout::InlineItemsBuilder::computeInlineTextItemWidths):
* Source/WebCore/layout/formattingContexts/inline/InlineLine.cpp:
(WebCore::Layout::Line::append):
* Source/WebCore/layout/formattingContexts/inline/IntrinsicWidthHandler.cpp:
(WebCore::Layout::isContentEligibleForNonLineBuilderMaximumWidth):
(WebCore::Layout::mayUseContentWidthBetweenLineBreaksAsMaximumSize):
(WebCore::Layout::IntrinsicWidthHandler::simplifiedMinimumWidth const):

Canonical link: <a href="https://commits.webkit.org/274729@main">https://commits.webkit.org/274729@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8746ac3876389e63872219386031de3fefd9a924

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39849 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18860 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42224 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42393 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35751 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42156 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21764 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16190 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33214 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40423 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15894 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/34456 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13754 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13779 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35412 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43672 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36184 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35748 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39516 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14664 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12062 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37790 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16283 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8939 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16331 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15927 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp:
```diff
@@ -67,8 +67,8 @@ static std::optional<InlineItemRange> partialRangeForDamage(const InlineItemList
-    auto& damagedInlineItem = inlineItemList[damageStartPosition.index];
-    if (damageStartPosition.offset && (!is<InlineTextItem>(damagedInlineItem) || damageStartPosition.offset >= downcast<InlineTextItem>(damagedInlineItem).length())) {
+    auto* damagedInlineItem = dynamicDowncast<InlineTextItem>(inlineItemList[damageStartPosition.index]);
+    if (damageStartPosition.offset && (!damagedInlineItem || damageStartPosition.offset >= damagedInlineItem->length())) {
```
I think this change makes the code less readable. `is<InlineTextItem>` is not strictly a prerequisite, but it expresses the fact that non-text type inline items cannot have offsets.

alanbaradlay commented on Source/WebCore/inspector/InspectorAuditAccessibilityObject.cpp:
```diff
@@ -68,7 +68,8 @@ ExceptionOr<Vector<Ref<Node>>> InspectorAuditAccessibilityObject::getElementsByC
     Vector<Ref<Node>> nodes;
 
-    for (Element& element : descendantsOfType<Element>(is<ContainerNode>(container) ? downcast<ContainerNode>(*container) : document)) {
+    auto* containerNode = dynamicDowncast<ContainerNode>(container);
+    for (Element& element : descendantsOfType<Element>(containerNode ? *containerNode : document)) {
```
not layout code.

alanbaradlay commented on Source/WebCore/inspector/InspectorAuditAccessibilityObject.cpp:
```diff
@@ -275,10 +276,8 @@ ExceptionOr<RefPtr<Node>> InspectorAuditAccessibilityObject::getMouseEventNode(N
-        if (is<AccessibilityNodeObject>(axObject))
-            return downcast<AccessibilityNodeObject>(axObject)->mouseButtonListener(MouseButtonListenerResultFilter::IncludeBodyElement);
-    }
+    if (auto* accessibilityNodeObject = dynamicDowncast<AccessibilityNodeObject>(accessibilityObjectForNode(node)))
+        return accessibilityNodeObject->mouseButtonListener(MouseButtonListenerResultFilter::IncludeBodyElement);
```
not layout code

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
```diff
@@ -160,8 +160,8 @@ bool InlineItemsBuilder::traverseUntilDamaged(LayoutQueue& layoutQueue, const Bo
-    if (!shouldSkipSubtree && is<ElementBox>(subtreeRoot) && downcast<ElementBox>(subtreeRoot).hasChild()) {
-        auto& firstChild = *downcast<ElementBox>(subtreeRoot).firstChild();
+    if (auto* elementBox = dynamicDowncast<ElementBox>(subtreeRoot); elementBox && elementBox->hasChild() && !shouldSkipSubtree) {
+        auto& firstChild = *elementBox->firstChild();
```
Now we unnecessarily call dynamicDowncast on plenty `shouldSkipSubtree` containers. 😿

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
```diff
@@ -448,14 +448,15 @@ static inline void buildBidiParagraph(const RenderStyle& rootStyle, const Inline
         auto& inlineItem = inlineItemList[index];
         auto& layoutBox = inlineItem.layoutBox();
-        auto mayAppendTextContentAsOneEntry = is<InlineTextBox>(layoutBox) && !TextUtil::shouldPreserveNewline(downcast<InlineTextBox>(layoutBox));
+        auto* inlineTextBox = dynamicDowncast<InlineTextBox>(layoutBox);
```
This introduces inlineTextBox on a scope where it should not really exist (see all the if else ifs further down)

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
```diff
@@ -635,7 +636,8 @@ void InlineItemsBuilder::breakAndComputeBidiLevels(InlineItemList& inlineItemLis
             // Mark the inline box stack with "content yes", when we come across a content type of inline item.
-            if (!inlineItem.isText() || !downcast<InlineTextItem>(inlineItem).isWhitespace() || TextUtil::shouldPreserveSpacesAndTabs(inlineItem.layoutBox()))
+            auto* inlineTextItem = dynamicDowncast<InlineTextItem>(inlineItem);
+            if (!inlineTextItem || !inlineTextItem->isWhitespace() || TextUtil::shouldPreserveSpacesAndTabs(inlineItem.layoutBox()))
```
we should probably stay consistent here.

`if (!inlineTextItem || !inlineTextItem->isWhitespace() || TextUtil::shouldPreserveSpacesAndTabs(inlineTextItem->layoutBox()))`

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/IntrinsicWidthHandler.cpp:
```diff
@@ -44,7 +44,11 @@ static bool isBoxEligibleForNonLineBuilderMinimumWidth(const ElementBox& box)
 static bool isContentEligibleForNonLineBuilderMaximumWidth(const ElementBox& rootBox, const InlineItemList& inlineItemList)
 {
-    return inlineItemList.size() == 1 && inlineItemList[0].isText() && !downcast<InlineTextItem>(inlineItemList[0]).isWhitespace() && rootBox.style().textIndent() == RenderStyle::initialTextIndent();
+    if (inlineItemList.size() != 1 || rootBox.style().textIndent() != RenderStyle::initialTextIndent())
+        return false;
+
+    auto* inlineTextItem = dynamicDowncast<InlineTextItem>(inlineItemList[0]);
+    return inlineTextItem && !inlineTextItem->isWhitespace();
```
this changes performance characteristic, but probably fine.

alanbaradlay commented on Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
```diff
@@ -159,13 +159,14 @@ bool InlineItemsBuilder::traverseUntilDamaged(LayoutQueue& layoutQueue, const Bo
-        if (traverseUntilDamaged(layoutQueue, firstChild, firstDamagedLayoutBox))
-            return true;
-        layoutQueue.takeLast();
+    if (!subtreeRoot.establishesFormattingContext()) {
```
any reason for removing `shouldSkipSubtree`?. it serves commentary purpose and makes the code understand better.

annevk committed:
More dynamicDowncast<> adoption in layout code
https://bugs.webkit.org/show_bug.cgi?id=269173

Reviewed by Alan Baradlay.

Including a drive-by fix for InspectorAuditAccessibilityObject.

* Source/WebCore/inspector/InspectorAuditAccessibilityObject.cpp:
(WebCore::InspectorAuditAccessibilityObject::getElementsByComputedRole):
(WebCore::InspectorAuditAccessibilityObject::getMouseEventNode):
* Source/WebCore/layout/formattingContexts/inline/InlineFormattingContext.cpp:
(WebCore::Layout::partialRangeForDamage):
(WebCore::Layout::isEmptyInlineContent):
* Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
(WebCore::Layout::requiresVisualReordering):
(WebCore::Layout::InlineItemsBuilder::traverseUntilDamaged):
(WebCore::Layout::InlineItemsBuilder::collectInlineItems):
(WebCore::Layout::InlineItemsBuilder::breakAndComputeBidiLevels):
(WebCore::Layout::InlineItemsBuilder::computeInlineTextItemWidths):
* Source/WebCore/layout/formattingContexts/inline/InlineLine.cpp:
(WebCore::Layout::Line::append):
* Source/WebCore/layout/formattingContexts/inline/IntrinsicWidthHandler.cpp:
(WebCore::Layout::isContentEligibleForNonLineBuilderMaximumWidth):
(WebCore::Layout::mayUseContentWidthBetweenLineBreaksAsMaximumSize):
(WebCore::Layout::IntrinsicWidthHandler::simplifiedMinimumWidth const):

Canonical link: https://commits.webkit.org/274729@main
